synapse/handlers/space_summary.py

```diff
@@ -76,67 +78,154 @@ async def get_space_summary(
         await self._auth.check_user_in_room_or_world_readable(room_id, requester)
 
         # the queue of rooms to process
-        room_queue = deque((room_id,))
+        room_queue = deque((_RoomQueueEntry(room_id),))
 
         processed_rooms = set()  # type: Set[str]
 
         rooms_result = []  # type: List[JsonDict]
         events_result = []  # type: List[JsonDict]
 
-        now = self._clock.time_msec()
+        while room_queue and len(rooms_result) < MAX_ROOMS:
+            queue_entry = room_queue.popleft()
+            room_id = queue_entry.room_id
+            logger.debug("Processing room %s", room_id)
+            processed_rooms.add(room_id)
+
+            # The client-specified max_rooms_per_space limit doesn't apply to the
+            # room_id specified in the request, so we ignore it if this is the
+            # first room we are processing.
+            max_children = max_rooms_per_space if processed_rooms else None
+
+            rooms, events = await self._summarize_local_room(
+                requester, room_id, suggested_only, max_children
+            )
+
+            rooms_result.extend(rooms)
+            events_result.extend(events)
+
+            # add any children that we haven't already processed to the queue
+            for edge_event in events:
+                if edge_event["state_key"] not in processed_rooms:
+                    room_queue.append(_RoomQueueEntry(edge_event["state_key"]))
+
+        return {"rooms": rooms_result, "events": events_result}
+
+    async def federation_space_summary(
+        self,
+        room_id: str,
+        suggested_only: bool,
+        max_rooms_per_space: Optional[int],
+        exclude_rooms: Iterable[str],
+    ) -> JsonDict:
```

Comment:
Looks like this is pretty much identical to the C-S implementation, but it:

* Doesn't take a requester (and thus doesn't check if the user is in the room, falling back to whether the room is world-readable instead).
* Doesn't use `_RoomQueueEntry`.
* Falls back to the default for the maximum number of rooms per entry.

Am I missing anything else?

Reply:
That sounds about right so far. The differences get a bit bigger in #9653: it also doesn't ever call out over federation, which means it doesn't need to track `via`s (which is why it doesn't use `_RoomQueueEntry`.

Reply:
Got it. Thanks! I was wondering why it was separate, but that makes sense that #9653 would have this differ more.